normandy/recipes/utils.py

```diff
@@ -46,3 +54,68 @@ def deterministic_sample(rate, inputs):
     assert len(input_hash) == 64
 
     return input_hash < sample_point
+
+
+class Autographer:
+    """
+    Interacts with an Autograph service.
+
+    If Autograph signing is not configured using `settings.AUTOGRAPH`,
+    raises `ImproperlyConfigured`. If the Autograph server can't be reached
+    or returns an HTTP error, an error will be thrown by `requests`.
+    """
+
+    def __init__(self):
+        self.check_config()
+        self._session = None
+
+    @property
+    def session(self):
+        if self._session is None:
+            self._session = requests.Session()
+            self._session.auth = HawkAuth(
+                id=str(settings.AUTOGRAPH_HAWK_ID),
+                key=str(settings.AUTOGRAPH_HAWK_SECRET_KEY))
+        return self._session
+
+    def check_config(self):
+        required_keys = ['URL', 'HAWK_ID', 'HAWK_SECRET_KEY']
+        for key in required_keys:
+            if getattr(settings, 'AUTOGRAPH_' + key) is None:
+                msg = 'set settings.AUTOGRAPH_{} to use action signatures'.format(key)
+                raise ImproperlyConfigured(msg)
+
+    def sign_data(self, content_list):
+        """
+        Fetches Signatures objects from Autograph for each item in `content_list`.
+
+        The items in `content_list` must be bytes objects.
+        """
+        ts = timezone.now()
+        url = '{}sign/data'.format(settings.AUTOGRAPH_URL)
+        signing_request = []
+        for item in content_list:
+            # base64 works in bytes. requests work in UTF-8.
+            # Convert to bytes, and then back.
+            encoded_implementation = base64.b64encode(item).decode('utf8')
+            signing_request.append({
+                'template': 'content-signature',
+                'input': encoded_implementation,
+            })
+
+        res = self.session.post(url, json=signing_request)
+        res.raise_for_status()
+        signing_responses = res.json()
+
+        from normandy.recipes.models import Signature  # avoid circular import
+        signatures = []
+        for res in signing_responses:
+            sig = Signature(
```

Comment:
I think the creation of this is complex enough that it *should* be a model. Returning a dict just makes callers do something like `[Signature(**sig_data) for sig_data in autographer.sign_data()]`, which doesn't seem better to me.

The key part for me is that I don't see Autographer being used for anything except generating Signature instances. Maybe it should be a part of the Signature model?